dot/state/pruner.go

```diff
@@ -0,0 +1,274 @@
+package state
+
+import (
+	"fmt"
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/scale"
+)
+
+const (
+	journalPrefix = "journal"
+	lastPrunedKey = "last_pruned"
+)
+
+// Pruner is implemented by FullNodePruner and ArchivalNodePruner.
+type Pruner interface {
+	StoreJournalRecord(deleted, inserted []*common.Hash, blockHash *common.Hash, blockNum *big.Int) error
+}
+
+// ArchivalNodePruner is a no-op since we don't prune nodes in archive mode.
+type ArchivalNodePruner struct{}
+
+// StoreJournalRecord for archive node doesn't do anything.
+func (a *ArchivalNodePruner) StoreJournalRecord(deleted, inserted []*common.Hash, blockHash *common.Hash, blockNum *big.Int) error {
+	return nil
+}
+
+// FullNodePruner stores state trie diff and allows online state trie pruning
+type FullNodePruner struct {
+	deathList     []*deathRow
+	storageDB     chaindb.Database
+	journalDB     chaindb.Database
+	deathIndex    map[*common.Hash]int64
+	pendingNumber int64
```

Comment:
please add a comment explaining that this is the next block to be pruned

Reply:
Done